### ISSUE TITLE
[feat] Implement CART (Context-Adaptive Token-Level Noise Rescheduling) for Dream-style training

### ISSUE THREAD

**Issue by nishide-dev** [nishide-dev/unturtle#202]
## Background

Dream 7B's key training innovation is **CART (Context-Adaptive Token-Level Noise Rescheduling)**, which dynamically re-weights the diffusion loss per masked token based on its local context. Instead of applying a uniform noise level to all tokens, CART assigns higher loss weight to tokens that are farther from clean (unmasked) context, reflecting their higher prediction difficulty.

The weight matrix is based on a **geometric distribution over distances** between positions:

```python
def context_adaptive_reweight(seq_len, cart_p=0.8):
    position_ids_l = np.arange(seq_len).reshape(-1, 1)
    position_ids_r = np.arange(seq_len).reshape(1, -1)
    distance = torch.from_numpy(position_ids_l - position_ids_r)
    
    # Geometric decay: w(i,j) = 0.5 * p * (1-p)^(|i-j|-1)  for i≠j, else 0
    # Where p=cart_p controls locality sharpness
    ...
```

For each masked position `n`, the CART weight is:
```
w(n) = Σ_i  Geo(cart_p, |n-i| - 1)  for all clean (unmasked) positions i
```
This biases training toward positions with fewer clean neighbors, i.e., more informationally dense regions (math operators, variable declarations, etc.).

### Reference implementation
- `dev/repos/Dream/src/trainer/fsdp_sft_trainer.py` L91–115: `context_adaptive_reweight()`
- L805–821: weight matrix applied as:
  ```python
  weight = non_mask.matmul(weight_matrix).masked_fill(non_mask, 0)
  loss = loss * weight.reshape(-1)
  ```
- Also available via `time_reweighting="cart"` in the Dream config

### Current state
`DiffusionTrainer` supports `loss_weight_type="uniform" | "timestep" | "scheduler"` but has no CART option. Users cannot replicate Dream's training quality without this.

Note: CART weight is applied **after** per-token CE, separately from the alpha-scheduler weight. It does **not** replace timestep `t` weighting — it is an additional position-based reweighting applied on top.

## Acceptance criteria

- [ ] Add `loss_weight_type="cart"` to `DiffusionTrainingArguments` with `cart_p: float = 0.8`
- [ ] Implement `context_adaptive_reweight(seq_len, cart_p)` in `unturtle/diffusion/schedulers.py` or a new `unturtle/diffusion/reweighting.py`
- [ ] Apply CART weight in `DiffusionTrainer._build_loss_weights` when `loss_weight_type="cart"`:
  - Weight matrix is `(L, L)` and can be pre-computed once per batch if seq_len is fixed
  - Combine with diffusion mask: `weight = clean_mask_float @ weight_matrix`, masked at clean positions to 0
  - Pass resulting `(B, L)` weight tensor to `fused_masked_diffusion_loss`
- [ ] Add test comparing loss curves with/without CART on a synthetic task showing that CART concentrates loss on harder tokens
- [ ] Document the parameter `cart_p` and its effect on locality sharpness

## References
- Dream paper: https://arxiv.org/abs/2508.15487
- `dev/repos/Dream/src/trainer/fsdp_sft_trainer.py`

## Labels
type: feat, diffusion

**Comment by nishide-dev** [comment]
Implemented and merged in PR #207. CART (context_adaptive_reweight) added in unturtle/diffusion/reweighting.py; cart_p=1.0 NaN bug fixed before merge.
